### ISSUE TITLE
How to implement dynamic shape ops?

### ISSUE THREAD

**Issue by pixelspark** [webonnx/wonnx#35]
Several ONNX operators such as [Reshape](https://github.com/onnx/onnx/blob/main/docs/Operators.md#Reshape) and [ConstantOfShape](https://github.com/onnx/onnx/blob/main/docs/Operators.md#constantofshape) take in two inputs, where the first is the input data and the second defines the shape of the output data.

This means that (at least in theory) the shape can be dynamic and so is the shape of the output of the node. This would mean that we cannot compile shaders for the next ops up front because the input shape might have changed (and some operators depend on the input shape).  Below is an example from the 'Tiny YOLO v3' model that shows dynamic reshaping:

![image](https://user-images.githubusercontent.com/181337/148902857-83842d5a-bc3f-4840-800d-ac1ae7fc9f7a.png)

We should think about whether to implement this (necessary if we want to support models like YOLO, BERT) but also how. Possibly the 'shape' parts can be calculated in advance (e.g. before running the meat of the model) but in some cases it is possibly so dynamic that we'd have to compile shaders during runtime (which we'd rather not do because of the performance impact). Would love to your hear thoughts!

**Comment by haixuanTao** [comment]
I see.

So I checked the tiny-yolov3 and it seems that the data used for defining the shape of Reshaping is initialized a little bit earlier and when passed through onnx-simplifier, the reshape shape is fixed making it possible to do everything at compile time. I think that BERT can also have all its tensor size static. I believe they have left those ops here for training purposes.

The command for simplifying tiny yolov3 is: `python -m onnxsim tiny-yolov3-11.onnx opt-tiny-yolov3-11.onnx --input-shape input_1:1,3,416,416 image_shape:1,2`

On a larger note, I believe that being able to have more dynamically shaped tensors is a good thing. But I think that it is going to be quite a stretch within the webgpu framework limits. And there is a good chance that having more flexibility comes at a performance cost. e.g. Having to read GPU memory mid computation. 

If this were to be implemented, I would prefer to have a concrete case example, so that we know what to look for. 

Also happy to hear your thought based on this :)

**Comment by pixelspark** [comment]
If the dynamism is not needed for inference, and can be statically removed by tools like onnx-simplifier, I am fine with not supporting them! We should probably check some popular models we'd want to support.

(I didn't know onnx-simplifier was also capable of removing these ops. I am on an Apple M1 and onnx-simplifier fails to install here, so I used the onnx Python library directly to do the dimensions inference. I will check if `tract` supports the inference+simplification and perhaps make a Rust program to run it on models for those who can't run onnx-simplifier. After all the whole point of WONNX is that you don't have to depend on Python anymore :-)).

**Comment by haixuanTao** [comment]
Oh... I didn't know that onnx-simplifier did not work on mac. 

Yes, I think that a compiling library in the liking of onnx-simplifier could be very beneficial ! And it could also compile other formats such as tensorflow pb, etc... 

I think that optimisation should be done as much as possible outside of this crate as it will reduce the size of the lib and its dependencies, which is great for a runtime. If you do so, I will contribute as much as possible.

**Comment by pixelspark** [comment]
> Oh... I didn't know that onnx-simplifier did not work on mac.

Yeah well it works on Mac, just not on the new shiny ones :-) Apparently onnx-simplifier depends on onnxruntime, which is not (yet?) available.

> I think that optimisation should be done as much as possible outside of this crate as it will reduce the size of the lib and its dependencies, which is great for a runtime. If you do so, I will contribute as much as possible.

For this you would probably need to turn the ONNX model into some kind of tree structure, and then walk that tree to remove or collapse certain branches. The same tree could then be used for efficient model execution as well (streamlines all the various 'buffer_infos' maps as well) so it might still be worthwhile to have it in the library. Once you have the tree the optimization itself doesn't have to be that heavy; it basically means doing these things:

* Statically inferring dimensions (this you can do quite easily i presume by creating a function/enum that knows what an op will produce given certain input dimensions)
* Constant folding (i.e. removing sequences of ops that will just compute the same value each time). You are sort of already doing a crude form of this for e.g. Split/Resize, by replacing inputs fed by 'initializers' with attributes.
* Inferring 'dynamic' dimensions (i.e. defined by the above sequences)

Tract is apparently taking a similar approach with their 'HIR' high-level intermediate representation (which they based on NNEF, which is basically an ONNX-like format but much simpler, and only intended for inference)

**Comment by pixelspark** [comment]
This is much easier to fix after #45 lands!

**Comment by zimond** [comment]
OK... I'm hitting this problem after hours of debug. 😅😅 Wish there would be a fix soon.

**Comment by zimond** [comment]
I get it, just need to enrich optimizer into another large match and fill output_shapes for each op_def
